[WebXR][OpenXR] Add API to get consent required/optional features of WebKitXRPermissionRequest

fujii opened the pull request:
#### 91c148a27b2ec626f3f235606bf8164e1d81d42f
<pre>
[WebXR][OpenXR] Add API to get consent required/optional features of WebKitXRPermissionRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=298178">https://bugs.webkit.org/show_bug.cgi?id=298178</a>

Reviewed by Adrian Perez de Castro.

MiniBrowser grant all optional features at the moment. GTK MiniBrowser
should have a UI to select optional features to grant.

Needs to add API tests.

* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
(toFeatureList):
(webkitXRPermissionRequestAllow):
(webkit_xr_permission_request_get_granted_features):
(webkit_xr_permission_request_get_consent_required_features):
(webkit_xr_permission_request_get_consent_optional_features):
(webkit_xr_permission_request_get_required_features_requested):
(webkit_xr_permission_request_get_optional_features_requested):
(webkit_xr_permission_request_set_granted_features):
(toWebKitXRSessionFeatures):
(webkitXRPermissionRequestCreate):
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequestPrivate.h:
* Tools/MiniBrowser/gtk/BrowserTab.c:
(WebKitXRSessionFeaturesToString):
(decidePermissionRequest):
* Tools/MiniBrowser/wpe/main.cpp:
(decidePermissionRequest):

Canonical link: <a href="https://commits.webkit.org/299591@main">https://commits.webkit.org/299591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66dedc7ad4234426a3c6d9e3cf51c75b375c6787

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38654 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125174 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71036 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47236 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90286 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59794 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31373 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106666 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70792 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30436 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24780 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68826 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111091 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100819 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24968 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128213 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45880 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98952 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/118397 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46247 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102886 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98732 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44191 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22191 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42449 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19016 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45750 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51428 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146185 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45216 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/146185 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48561 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46901 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos reviewed (commented):
We need API tests for this too.

fujii commented:
> We need API tests for this too.

Simon added a test case in [this PR](https://github.com/WebKit/WebKit/pull/49055), but removed.
And, created a bug ticket for it.  https://bugs.webkit.org/show_bug.cgi?id=297261
I should wait for it.

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+static WebKitXRSessionMode
+toWebKitXRSessionMode(PlatformXR::SessionMode mode)
```
Fixed.

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+static WebKitXRSessionFeatures
+toWebKitXRSessionFeatures(const PlatformXR::Device::FeatureList& features)
```
Fixed.

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
```diff
@@ -0,0 +1,99 @@
+webkit_xr_permission_request_get_optional_features (WebKitXRPermissionRequest *request);
+
+WEBKIT_API void
+webkit_xr_permission_request_set_granted_features (WebKitXRPermissionRequest *request, WebKitXRSessionFeatures granted);
```
Fixed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cd6882c97e6b321a1197694eb6df76cff906b804. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49447

fujii commented on Tools/MiniBrowser/gtk/BrowserTab.c:
```diff
@@ -398,6 +398,17 @@ static gboolean decidePermissionRequest(WebKitWebView *webView, WebKitPermission
+        WebKitXRSessionFeatures requiredFeatures = webkit_xr_permission_request_get_required_features(xrRequest);
+        WebKitXRSessionFeatures optionalFeatures = webkit_xr_permission_request_get_optional_features(xrRequest);
+        gchar *originStr = webkit_security_origin_to_string(origin);
+        text = g_strdup_printf("Allow XR device access?\norigin=%s mode=%d requiredFeatures=%x optionalFeatures=%x", originStr, mode, requiredFeatures, optionalFeatures);
```
Fixed.

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
```diff
@@ -0,0 +1,258 @@
+ *
+ * Get the origin of this request.
+ *
+ * Returns: (transfer none): A #WebKitSecurityOrigin of @request
```
Basically, no. But, it returns NULL only if the argument `request` is nullptr.

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
```diff
@@ -0,0 +1,258 @@
+    return static_cast<WebKitXRSessionFeatures>(result);
+}
+
+WebKitXRPermissionRequest* webkitXRPermissionRequestCreate(const WebCore::SecurityOriginData& securityOriginData, PlatformXR::SessionMode mode, const PlatformXR::Device::FeatureList& /* granted */, const PlatformXR::Device::FeatureList& /* consentRequired */, const PlatformXR::Device::FeatureList& /* consentOptional */, const PlatformXR::Device::FeatureList& requiredFeaturesRequested, const Platf ...
```
I also wonder. Will check.

fujii replied:
I've read <https://immersive-web.github.io/webxr/#permissions>. I was completely wrong. I should pass `granted`, `consentRequired` and `consentOptional` to the user.

spenap commented:
> I should wait for it.

https://github.com/WebKit/WebKit/pull/49322 has the test. It is currently a draft, rebased against @carlosgcampos 's https://github.com/WebKit/WebKit/pull/49448, as his PR reworked the render thread into a WorkQueue

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
```diff
@@ -0,0 +1,109 @@
+webkit_xr_permission_request_get_optional_features_requested (WebKitXRPermissionRequest *request);
+
+WEBKIT_API void
+webkit_xr_permission_request_set_granted_features            (WebKitXRPermissionRequest *request,
```
The member name of XRPermissionStatus is `granted`. https://immersive-web.github.io/webxr/#resolve-the-requested-features
Should I rename webkit_xr_permission_request_set_granted_features to webkit_xr_permission_request_set_granted?

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
```diff
@@ -0,0 +1,109 @@
+webkit_xr_permission_request_get_session_mode                (WebKitXRPermissionRequest *request);
+
+WEBKIT_API WebKitXRSessionFeatures
+webkit_xr_permission_request_get_granted_features            (WebKitXRPermissionRequest *request);
```
We have two API `webkit_xr_permission_request_{get,set}_granted_features`. they are irrelevant.
Should I rename webkit_xr_permission_request_get_granted_features to webkit_xr_permission_request_get_previously_granted_features?

fujii commented:
<img width="1646" height="1036" alt="Screenshot From 2025-08-19 10-44-03" src="https://github.com/user-attachments/assets/e3d63659-9ba2-4cda-b341-441c8c07831e" />

The demo content requests "viewer" and "local" features. But, since [the device already enables the features](https://github.com/WebKit/WebKit/blob/a8e27303bb22c49a8dd08bc8c96d85e64135d296/Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp#L96-L100), they are already `granted`. So, the permission dialog is shown with consentRequired and consentOptional are empty. Is this the expected behavior?

fujii commented:
API with the session features seem to need more discussions. I'm going to propose a simpler API as the first step in #49631.

fujii commented on Tools/MiniBrowser/gtk/BrowserTab.c:
```diff
@@ -419,8 +438,20 @@ static gboolean decidePermissionRequest(WebKitWebView *webView, WebKitPermission
+            "origin=%s mode=%s granted=(%s) consentRequired=(%s) consentOptional=(%s) requiredFeaturesRequested=(%s) optionalFeaturesRequested=(%s)",
+            originStr, webKitXRSessionModeToString(mode), grantedFeatures, consentRequiredFeatures, consentOptionalFeatures, requiredFeaturesRequested, optionalFeaturesRequested);
         g_free(originStr);
+        g_free(grantedFeatures);
+        g_free(consentRequiredFeatures);
+        g_free(consentOptionalFeatures);
+        g_free(requiredFeaturesRequested);
+        g_free(optionalFeaturesRequested);
```
Sounds an interesting idea. We can remove lots of `g_free`. But, I don't want to include a refactoring of existing code to this change. I will try it if I have a chance.

fujii replied:
Fixed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/66dedc7ad4234426a3c6d9e3cf51c75b375c6787. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49447

fujii commented:
> I don't think we should show it in that case, i.e., if both are empty

Oh, it's a stale comment. https://commits.webkit.org/299314@main turned on ENABLE_WEBXR_HANDS. Now, `hand-tracking` is the only explicitly consent required features.

fujii commented:
Thank you very much for the review. Simon is working on the API test in https://github.com/WebKit/WebKit/pull/49322 . I will add a test after that.

fujii committed:
[WebXR][OpenXR] Add API to get consent required/optional features of WebKitXRPermissionRequest
https://bugs.webkit.org/show_bug.cgi?id=298178

Reviewed by Adrian Perez de Castro.

MiniBrowser grant all optional features at the moment. GTK MiniBrowser
should have a UI to select optional features to grant.

Needs to add API tests.

* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.cpp:
(toFeatureList):
(webkitXRPermissionRequestAllow):
(webkit_xr_permission_request_get_granted_features):
(webkit_xr_permission_request_get_consent_required_features):
(webkit_xr_permission_request_get_consent_optional_features):
(webkit_xr_permission_request_get_required_features_requested):
(webkit_xr_permission_request_get_optional_features_requested):
(webkit_xr_permission_request_set_granted_features):
(toWebKitXRSessionFeatures):
(webkitXRPermissionRequestCreate):
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitXRPermissionRequestPrivate.h:
* Tools/MiniBrowser/gtk/BrowserTab.c:
(WebKitXRSessionFeaturesToString):
(decidePermissionRequest):
* Tools/MiniBrowser/wpe/main.cpp:
(decidePermissionRequest):

Canonical link: https://commits.webkit.org/299591@main

fujii commented on Tools/MiniBrowser/gtk/BrowserTab.c:
```diff
@@ -341,6 +341,27 @@ static const gchar *webKitXRSessionModeToString(WebKitXRSessionMode mode)
+    unsigned count = 0;
+    GString *str = g_string_new(NULL);
+    if (features & WEBKIT_XR_SESSION_FEATURES_VIEWER)
+        g_string_append(str, " viewer" + !(count++));
```
Fixed. https://commits.webkit.org/299670@main
